# BUSA3020 - Assignment 2  

**Assignment Points**: 100  
**Assignment Weight**: 20%  
**Submission**: Provide your answers in this notebook and submit it via iLearn link

- Do NOT use `print()` in the final version of the assignment unless you are explicitly asked to do so
- Make sure your answers are direct and easy to read
- Marks will be deducted for not following instructions

---

**Problem 1** - Reading and describing data (Total Marks: 20)



**Q1**. Read the **modified** credit card dataset provided in the **assignment_data** folder contained in the assignment zip file
- Name your DataFrame `df`  
- Rename the columns 'PAY_0' and 'default payment next month' as in Programming Task 1 
- Delete 'ID' column  

(5 marks) 




In [1]:
import pandas as pd

pd.set_option("display.max_rows", None, "display.max_columns", None, "display.width", None)
df = pd.read_excel('assignment_data/credit_card_defaults.xlsx')

df.rename(columns={'default payment next month': 'payment_default'}, inplace = True)
df.rename(columns={'PAY_0': 'PAY_1'}, inplace = True)

df.drop("ID", axis=1, inplace=True)
df.head()

Unnamed: 0  LIMIT_BAL  SEX  EDUCATION  MARRIAGE   AGE  PAY_1  PAY_2  PAY_3  \
0           0    20000.0  2.0        2.0       1.0  24.0    2.0    2.0   -1.0   
1           1   120000.0  2.0        2.0       2.0  26.0   -1.0    2.0    0.0   
2           2    90000.0  2.0        2.0       2.0  34.0    0.0    0.0    0.0   
3           3    50000.0  2.0        2.0       1.0  37.0    0.0    0.0    0.0   
4           4    50000.0  1.0        2.0       1.0  57.0   -1.0    0.0   -1.0   

   PAY_4  PAY_5  PAY_6  BILL_AMT1  BILL_AMT2  BILL_AMT3  BILL_AMT4  BILL_AMT5  \
0   -1.0   -2.0   -2.0     3913.0     3102.0      689.0        0.0        0.0   
1    0.0    0.0    2.0     2682.0     1725.0     2682.0     3272.0     3455.0   
2    0.0    0.0    0.0    29239.0    14027.0    13559.0    14331.0    14948.0   
3    0.0    0.0    0.0    46990.0    48233.0    49291.0    28314.0    28959.0   
4    0.0    0.0    0.0     8617.0     5670.0    35835.0    20940.0    19146.0   

   BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0        0.0       0.0     689.0       0.0       0.0       0.0         0   
1     3261.0       0.0    1000.0    1000.0    1000.0       0.0      2000   
2    15549.0    1518.0    1500.0    1000.0    1000.0    1000.0      5000   
3    29547.0    2000.0    2019.0    1200.0    1100.0    1069.0      1000   
4    19131.0    2000.0   36681.0   10000.0    9000.0     689.0       679   

   payment_default  
0                1  
1                1  
2                0  
3                0  
4                0

<hr style="width:25%;margin-left:0;"> 

**Q2**. Explain which features are numeric, ordinal, and nominal variables, and how many features of each variable type there are in the dataset
- To answer this question look at the data itself and consider the definitions of the variables (provided elsewhere in course material)
- Your final answer should consist of text only and not contain any python code

(10 marks)

- **LIMIT_BAL** is a numerical variable, with minimum at 10,000 to maximum of 1,000,000
- **SEX** is a nominal variable, where 1 = male and 2 = female (2 features)  
- **EDUCATION** is a nominal variable, where the values 1-6 represent the level of education (1 = graduate school, 3 = high-school, 2 = University, 4 = Other, 5 = unknown, 6 = unknown) (6 features)  
- **MARRIAGE** is a nominal variable, represented by 4 values (1 = married, 2 = single, 3 = other, 4 = unknown)    
- **AGE** is a discrete numeric variable
- **PAY_1-6** is a ordinal variable, representing repayment status in months 
- **BILL_AMOUNT 1-6** is a continuous numerical variable, representing bill statement in dollars  
- **PAY_AMOUNT1-6** is a continuous numerical variable, representing previous payment in dollars    
- **PAYMENT_DEFAULT** is a nominal variable where 1 = yes and 0 = no (2 features)  

**Numerical** - 14
**Ordinal** - 6
**Nominal** - 4

<hr style="width:25%;margin-left:0;"> 

**Q3.** Describing the dataset. 

- Print `info()` and comment on the missing values in the dataset

(5 marks)

In [2]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10001 entries, 0 to 10000
Data columns (total 25 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       10001 non-null  int64  
 1   LIMIT_BAL        9993 non-null   float64
 2   SEX              9993 non-null   float64
 3   EDUCATION        9983 non-null   float64
 4   MARRIAGE         9983 non-null   float64
 5   AGE              9983 non-null   float64
 6   PAY_1            9983 non-null   float64
 7   PAY_2            9983 non-null   float64
 8   PAY_3            9971 non-null   float64
 9   PAY_4            9964 non-null   float64
 10  PAY_5            9964 non-null   float64
 11  PAY_6            9971 non-null   float64
 12  BILL_AMT1        9971 non-null   float64
 13  BILL_AMT2        9979 non-null   float64
 14  BILL_AMT3        9979 non-null   float64
 15  BILL_AMT4        9979 non-null   float64
 16  BILL_AMT5        9979 non-null   float64
 17  BILL_AMT6   

Looking at the data, as well as the count of columns which can be represented in column 1 (unnamed) which acts as an index, we can see there are 10001 rows altogether. It also states that there are 10001 entries. Looking at each column from LIMIT_BAL to PAY_AMT5, we can see there are missing values as it does not equal to the number of rows. Alternatively, we can see that PAY_AMT6 and payment_default both have 10001 which means there are no missing values for those columns

---
---

**Problem 2.** Cleaning data and dealing with categorical features (Total Marks: 40)


**Q1.** 
- Use an appropriate `pandas` function to impute missing values (10 marks)
    - Be careful when deciding which method to use to replace missing observations 
    - Take into consideration the type of each variable and the best practices we discussed in class/lecture notes
- Explain what you have done and why (5 marks)

(Total: 15 marks)

In [3]:
# Mode
df['SEX'].fillna(df['SEX'].mode()[0], inplace=True)
df['EDUCATION'].fillna(df['EDUCATION'].mode()[0], inplace=True)
df['MARRIAGE'].fillna(df['MARRIAGE'].mode()[0], inplace=True) 
df['PAY_1'].fillna(df['PAY_1'].mode()[0], inplace=True)
df['PAY_2'].fillna(df['PAY_2'].mode()[0], inplace=True)
df['PAY_3'].fillna(df['PAY_3'].mode()[0], inplace=True)
df['PAY_4'].fillna(df['PAY_4'].mode()[0], inplace=True)
df['PAY_5'].fillna(df['PAY_5'].mode()[0], inplace=True)
df['PAY_6'].fillna(df['PAY_6'].mode()[0], inplace=True)

# Mean
df['AGE'].fillna(df['AGE'].mean(), inplace=True) 
df['LIMIT_BAL'].fillna(df['LIMIT_BAL'].mean(), inplace=True)
df['BILL_AMT1'].fillna(df['BILL_AMT1'].mean(), inplace=True)
df['BILL_AMT2'].fillna(df['BILL_AMT2'].mean(), inplace=True)
df['BILL_AMT3'].fillna(df['BILL_AMT3'].mean(), inplace=True)
df['BILL_AMT4'].fillna(df['BILL_AMT4'].mean(), inplace=True)
df['BILL_AMT5'].fillna(df['BILL_AMT5'].mean(), inplace=True)
df['BILL_AMT6'].fillna(df['BILL_AMT6'].mean(), inplace=True)
df['PAY_AMT1'].fillna(df['PAY_AMT1'].mean(), inplace=True)
df['PAY_AMT2'].fillna(df['PAY_AMT2'].mean(), inplace=True)
df['PAY_AMT3'].fillna(df['PAY_AMT3'].mean(), inplace=True)
df['PAY_AMT4'].fillna(df['PAY_AMT4'].mean(), inplace=True)
df['PAY_AMT5'].fillna(df['PAY_AMT5'].mean(), inplace=True)
df['PAY_AMT6'].fillna(df['PAY_AMT6'].mean(), inplace=True)

#print('df.isnull().sum()\n',df.isnull().sum())
#df.head()


As there is no repetition of the index that is trying to be replaced, there is no simpler way to replace all the the NaN values above. If there was, a loop function may have been used. 

For the data, either the column mode or mean was used for each column. The mode functionwas used for SEX, EDUCATION, MARRIAGE and PAY_1-6 as the numbers had an associated categorical label attached to it. The mode is the most common value in their respective column.

Furthermore, the mean function has been used for all numerical values being BILL_AMT1-6, PAY_AMT1-6, LIMIT_BAL and AGE which are not encoded to a categorical value and are simply infinitely large values. As our data does not showcase many or significant outliers, it is prefered to use the mean imputation method over the median. 


<hr style="width:25%;margin-left:0;"> 

**Q2**. 
- Print `value_counts()` of 'SEX' column and add dummy variables 'SEX_MALE' and 'SEX_FEMALE' to the `df` using `get_dummies()` (3 marks)
- Carefully explain how the dummy variables are constructed (1 mark)
- Delete the variable 'SEX' from `df` (1 mark)

(Total: 5 marks)  

In [4]:
#Value count
print(df.value_counts("SEX"))

#Dummy Variable creation
df['SEX'] = df['SEX'].apply(str)
SEX_dummies = pd.get_dummies(df['SEX'])
df = df.join(SEX_dummies)
df.rename(columns = {"1.0" : "SEX_MALE", "2.0" : "SEX_FEMALE"},inplace = True)

#Delete SEX
del df["SEX"]


SEX
2.0    5799
1.0    4202
dtype: int64


Dummy variables indicates whether a separate categorical variable takes on a specific value.
The initial sex column was comprised of 1.0 for Males and 2.0 for Females. To ensure we can look at these independently, we create dummy variables which are two additional columns which take on a 0 or 1 for whether they were Male or Female. As the row can only be a male or female, the value will be 1 for the actual value and 0 for the alternative choice. Nevertheless, this ensures our future models dont think that FEMALE > MALE in which would result in accurate models.

<hr style="width:25%;margin-left:0;"> 

**Q3**. Print `value_counts()` of 'MARRIAGE' column and *carefully* comment on what you notice in relation to the definition of this variable. 

(Total: 5 marks) 

In [5]:
print(df['MARRIAGE'].value_counts())

2.0    5476
1.0    4383
3.0     125
0.0      17
Name: MARRIAGE, dtype: int64


Analysing the data, it is apparent that most records are individuals who are single with 5458 with slightly less people who are married at 4383 recorded values. A very small proportion of people selected 'Other' as an option at just 125. Finally we can see that 17 records were recorded with a value of 0 which is not described in the data. This may be mistyped responses, data that was lost in mannual transfer or fields that were chosen not to be filled. Moving forward, this data should be reallocated.

<hr style="width:25%;margin-left:0;"> 

**Q4**. 

- Apply `get_dummies()` to 'MARRIAGE' feature and add dummy variables 'MARRIAGE_MARRIED', 'MARRIAGE_SINGLE', 'MARRIAGE_OTHER' to `df`. (2.5 marks)   
- Carefully consider how to allocate all values of 'MARRIAGE' across the 3 newly created features (5 marks)   
    - Explain what decisions you had to make
- Delete 'MARRIAGE' from `df` (2.5 marks)   

(Total: 10 marks)   

In [6]:
#Replace 0 Value
df["MARRIAGE"].replace([0],3,inplace=True)

#Dummy variable creation
df['MARRIAGE'] = df['MARRIAGE'].apply(str)
MARRIAGE_dummies = pd.get_dummies(df['MARRIAGE'])
df = df.join(MARRIAGE_dummies)
df.rename(columns = {'1.0': 'MARRIAGE_MARRIED', '2.0': 'MARRIAGE_SINGLE', '3.0': 'MARRIAGE_OTHER'},inplace = True)

#Delete Marriage
del df["MARRIAGE"]

As there is only a small number of value 0.0, we can assume the user inputted a response that would fit under other but they gave a specific response and did not pick other. For this reason, I have decided to replace the 0s with 3, which will put it into the OTHER column. Therefore, it is more suitable to put it into a pre-existing column that better suits the data as other, in comparison to putting it into a column which we certainly know it doesnt belong to. By implementing this prior to the get_dummies function, it resulted in no unwanted column with 0s. Furthermore, the MARRIGE column is dropped automaticaly when it is converted to the dummy variables and therefore does not need to be deleted mannualy.

<hr style="width:25%;margin-left:0;"> 

**Q5**. In the column 'EDUCATION', convert the values {0, 4, 5, 6} into the value 4. 

(Total: 5 marks)  

In [7]:
df["EDUCATION"].replace([0,4,5,6],4, inplace = True)

---
---

**Problem 3** Preparing X and y arrays (Total Marks: 10)

**Q1**. 

- Create `y` from the first 10,000 observations of 'payment_default' column from `df` (2.5 marks)   
- Create `X`  from the first 10,000 observations of all the remaining features in `df` (2.5 marks)   

(Total: 5 Marks)

In [8]:
import numpy as np
y = df['payment_default'].iloc[:10000].values
X = df.drop(['payment_default'], axis = 1).iloc[:10000].values

<hr style="width:25%;margin-left:0;"> 

**Q2**. 

- Use an appropriate scikit-learn library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% train and 30% test datasets (2.5 marks) 
    - Set random_state to 4 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
- Standardise the data using `StandardScaler` library (2.5 marks)   

(Total: 5 marks) 

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 4, stratify = y)

sc.fit(X_train)

X_train_scaled = sc.transform(X_train)
X_test_scaled = sc.transform(X_test)

---
---

**Problem 4**. Support Vector Classifier and Accuracies (Total Marks: 30)


**Q1**. 

- Train a Support Vector Classifier on standardised data (5 marks)
    - Use `linear` kernel and set `random_state` to 3 (don't change any other parameters)
- Compute and print training and test dataset accuracies (5 marks)

(Total: 10 marks)   

In [10]:
from sklearn.svm import SVC

svm = SVC(kernel='linear',random_state=3)

svm.fit(X_train_scaled, y_train)
svm.fit(X_test_scaled, y_test)

print(f'Accuracy - Training dataset {svm.score(X_train_scaled, y_train):.4f}')
print(f'Accuracy - Test dataset {svm.score(X_test_scaled, y_test):.4f}')

Accuracy - Training dataset 0.7816
Accuracy - Test dataset 0.7843


<hr style="width:25%;margin-left:0;"> 

**Q2.**

- Extract 3 linear principal components from the standardised features using an appropriate `sklearn` library (5 marks)
- Train a Support Vector Classifier on the 3 principal components computed above (5 marks)   
    - Use `linear` kernel and set `random_state` to 3 (don't change any other parameters)
- Compute and print training and test dataset accuracies (5 marks)


(Total: 15 marks)  

In [11]:
from sklearn.decomposition import PCA

#PCA
svm_PCA = SVC(kernel='linear',random_state=3)
pca = PCA(n_components = 3)

#Transforming & Fitting
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

svm_PCA.fit(X_train_pca, y_train)

#Accuracy
print(f'Accuracy - Training dataset {svm_PCA.score(X_train_pca, y_train):.4f}')
print(f'Accuracy - Test dataset {svm_PCA.score(X_test_pca, y_test):.4f}')

Accuracy - Training dataset 0.7741
Accuracy - Test dataset 0.7740


<hr style="width:25%;margin-left:0;"> 

**Q3**. 

- Compare and comment on the computed accuracies from the last two questions above
    - Make comparisons both within and across the two questions

(Total: 5 marks)     


Looking at the original model that has no dimension reduction, we can see the training and test accuracies have values that are very close to one another with 0.7816 for training and 0.7843 for testing. As the accuracy between the training and testing set are similar (<1%), we can assume that the models are not overfitting the data.

Altnernatively, the SVM model which uses PCA (n = 3) as a dimensional reduction method contains a training and testing accuracy score of 0.7741 and 0.7740 respectively. As the accuracy between the training and testing set are similar, we can assume that the models are not overfitting the data.


When comparing the two models, we can see that by utilising dimensional reduction methods like PCA, in this case it does not improve the accuracy of our support vector machine. This may be as a result of creating components which do not capture the majority of the variance in the data, and therefore, there is important features that is being excluded from this model. This is due to the model not consideirng the variable/prediction target, and only focuses on the variance. Therefore, the features which may have large variance may have nothing to do with the prediction value or PAYMENT_DEFAULT. As a result, possibly more principal components should be used. Hence, we may assume that the second model has a worse bias-variance tradeoff in comparison to the original model. Consequently, we believe model 1 fits the data better and predicts the classifcation correctly more often.